src/compiler/crystal/semantic/abstract_def_checker.cr

```diff
@@ -264,19 +279,66 @@ class Crystal::AbstractDefChecker
     return false if r1 && !r2
     if r2 && r1 && r1 != r2
       # Check if a1.restriction is contravariant with a2.restriction
+      rt1 = nil
+      rt2 = nil
+
       begin
         rt1 = t1.lookup_type(r1, free_vars: free_vars1)
+      rescue Crystal::TypeException
+      end
+
+      begin
         rt2 = t2.lookup_type(r2, free_vars: free_vars2)
-        return false unless rt2.implements?(rt1)
       rescue Crystal::TypeException
-        # Ignore if we can't find a type (assume the method is implemented)
-        return true
+      end
+
+      if rt1 && rt2
+        # Both types resolved - check compatibility
+        return false unless rt2.implements?(rt1)
+      elsif !rt2
+        # Only implementation side couldn't be resolved
+        if !can_lookup?(r2, free_vars2, t2)
+          report_undefined_type(r2)
+        end
+      elsif !rt1
+        # Only abstract side couldn't be resolved
+        if !can_lookup?(r1, free_vars1, t1)
+          report_undefined_type(r1)
+        end
       end
     end
 
     true
   end
 
+  # Check if a restriction can potentially be looked up (is a free var or might be resolved later)
+  private def can_lookup?(restriction : ASTNode, free_vars, type : Type? = nil) : Bool
+    return true unless restriction.is_a?(Path)
+
+    # For single-name paths, check if it's a free variable
+    if name = restriction.single_name?
+      # Check if it's in free_vars (forall parameters)
+      return true if free_vars && free_vars.has_key?(name)
+
+      # Check if it's a type parameter of the enclosing generic type
+      if type && type.is_a?(GenericType)
+        return true if type.type_vars.includes?(name)
+      end
+
+      return false
+    end
+
+    return true if restriction.global?
```

Comment:
At this time all the types have been defined, so I don't really know what this is supposed to do. I would expect to see an error in either abstract def here:

```crystal
abstract class Foo
  abstract def foo(x : Int32)
  abstract def bar(x : ::A)
end

class Bar < Foo
  def foo(x : ::A); end
  def bar(x : Int32); end
end
```